crates/noirc_evaluator/src/ssa_refactor/abi_gen/mod.rs

```diff
@@ -13,8 +13,57 @@ use noirc_abi::{Abi, AbiParameter, FunctionSignature};
 /// This function returns the lengths ordered such as to correspond to the ordering used by the
 /// SSA representation. This allows the lengths to be consumed as array params are encountered in
 /// the SSA.
-pub(crate) fn collate_array_lengths(_abi_params: &[AbiParameter]) -> Vec<usize> {
-    Vec::new()
+pub(crate) fn collate_array_lengths(abi_params: &[AbiParameter]) -> Vec<usize> {
+    let mut acc = Vec::new();
+    for abi_param in abi_params {
+        collate_array_lengths_recursive(&mut acc, &abi_param.typ);
+    }
+    acc
+}
+
+/// The underlying recursive implementation of `collate_array_lengths`
+///
+/// This does a depth-first traversal of the abi until an array (or string) is encountered, at
+/// which point arrays are handled differently depending on the element type:
+/// - arrays of fields, integers or booleans produce an array of the specified length
+/// - arrays of structs produce an array of the specified length for each field of the flatten
+///   struct (which reflects a simplification made during monomorphization)
+fn collate_array_lengths_recursive(acc: &mut Vec<usize>, abi_type: &AbiType) {
+    match abi_type {
+        AbiType::Array { length, typ: elem_type } => {
+            match elem_type.as_ref() {
+                AbiType::Array { .. } => {
+                    unreachable!("2D arrays are not supported");
+                }
+                AbiType::Struct { .. } => {
+                    // monomorphization converts arrays of structs into an array per flattened
+                    // struct field.
+                    let array_count = elem_type.field_count();
+                    for _ in 0..array_count {
+                        acc.push(*length as usize);
```

Comment:
What is this meant to be doing?

Reply:
Question was mainly around why we are doing: `acc.push(*length as usize);` ie pushing the same length for each element type

Reply:
It's to mirror the monomorphization destructing of arrays of structs. ie. 
```
Foo { x: Field, y: Field }
foos: [Foo; 3]
```
effectively becomes
```
foos_x: [Field; 3]
foos_y: [Field; 3]
```

Reply:
I think maybe it's clearer this is happening in #1460:
https://github.com/noir-lang/noir/pull/1460/files#diff-2ed899985027bb1ff51f049bb416596f3bb915d0794ac27ff3bc36f6afd682a3R43